cabot/cabotapp/tests/scripts/test_concurrent_activity_counters.sh

```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+#
+# Test concurrent activity-counter modification requests to make sure they
+# are properly synchronized. If not, we'll get incorrect counts as requests
+# clobber each other during DB writes.
+
+set -o pipefail
+set -o nounset
+set -o errexit
+
+URL='http://localhost:5001/api/status-checks/activity-counters?id=1000'
+NUM_REQS=30
+SLEEP_SEC=4
+EXPECTED="\"counter.count\": $NUM_REQS"
+
+# Setup the sample check to work with
+echo "Setting up sample check"
+python manage.py shell < cabot/cabotapp/tests/scripts/setup_sample_check.py
+
+# Get the existing count
+echo "Existing count"
+curl -s "$URL"
+
+# Increment 50 times
+echo "Incrementing $NUM_REQS times in the background"
+for NUM in $(seq 1 $NUM_REQS)
+do
+  curl -s "$URL&action=incr" &
+done
+
+# Wait until everything is done
+echo "Sleeping for $SLEEP_SEC seconds"
+sleep $SLEEP_SEC
+
+# Check the final count
+echo "Getting final count"
+RESPONSE=$(curl -s "$URL")
+echo "$RESPONSE"
+ACTUAL=$(egrep -o '"counter.count": [0-9]+' <<< "$RESPONSE")
+
+if [[ "$ACTUAL" = "$EXPECTED" ]]
+then
+  echo "SUCCESS: got expected response '$EXPECTED'"
+else
+  echo "FAILURE: got response '$ACTUAL', expected '$EXPECTED'"
+  exit 1
+fi
```

Comment:
Is there no way to do this in python? Don't really like mixing python and bash in tests + requiring running multiple things to test

Reply:
We have [other api tests](https://github.com/Affirm/cabot/blob/master/cabot/cabotapp/tests/test_api.py) so it seems doable?

Reply:
@elainearbaugh (and @Aloshi) Yeah, I tried to do this in Python at first but didn't have any luck.  It seems like Django tests don't play well with multiple threads.  Just as an example:

```
class TestActivityCounterAPI(LocalTestCase):
    ....
    @staticmethod
    def make_request():
        url = '/api/status-checks/activity-counter?id=10102&action=incr'
        Client().get(url)

    def test_single_threaded_works(self):
        self._set_activity_counter(True, 1)
        self.assertEqual(StatusCheck.objects.get(id=10102).activity_counter.count, 1)
        TestActivityCounterAPI.make_request()
        self.assertEqual(StatusCheck.objects.get(id=10102).activity_counter.count, 2)

    def test_multi_threaded_fails(self):
        self._set_activity_counter(True, 1)
        self.assertEqual(StatusCheck.objects.get(id=10102).activity_counter.count, 1)
        t = threading.Thread(target=TestActivityCounterAPI.make_request)
        t.start()
        t.join()
        self.assertEqual(StatusCheck.objects.get(id=10102).activity_counter.count, 2)
```

The first test passes, but the second fails and doesn't seem to actually update the database:

```
FAIL [0.106s]: test_multi_threaded_fails (cabot.cabotapp.tests.test_api.TestActivityCounterAPI)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/code/cabot/cabotapp/tests/test_api.py", line 461, in test_multi_threaded_fails
    self.assertEqual(StatusCheck.objects.get(id=10102).activity_counter.count, 2)
AssertionError: 1 != 2
```

I also tried [this wrapper](https://www.caktusgroup.com/blog/2009/05/26/testing-django-views-for-concurrency-issues/) with similar results.  It also complains about ongoing connections during test teardown when running multiple (10 in my case) threads.

So... not sure what the solution is here.  I'm not able to get concurrent requests working in django test code, and not sure that's even possible.  And from reading about the [test client](https://docs.djangoproject.com/en/2.1/topics/testing/tools/), it seems like it's not intended to be used that way.

Anyway, I don't like having a bash script for testing this either, but can't get it to work any other way.

Reply:
I think I managed to get it to work [here](https://github.com/Affirm/cabot/pull/152/commits/806585a75d41ce438ea48a547811c012d75ce47a) (tested that it fails without the `select_for_update()` in views.py, and works with it). I think the issue was that `LocalTestCase` subclasses `TestCase`, which supposedly wraps the whole test in a transaction. Inheriting from `TransactionTestCase` instead made it start working (see [docs](https://docs.djangoproject.com/en/2.1/topics/testing/tools/#transactiontestcase) and [blog post](http://rahmonov.me/posts/testcase-vs-transactiontestcase-in-django/)). I also had to manually close the DB connection in the test threads (by calling `django.db.connection.close()`).